File: Assignment-2/sep12/c/ode.c
#include "rk4.h"

// You need to include the following files for printf() and for exp(),
// respectively.
#include <stdio.h>
#include <math.h> 

// Skeleton of a sample  ODE.C file

// MYODE is the name of your actual vector field.  It must have the same
// arguments in the same order as the prototype in rk4.h.

void myode(real_t t, size_t neq, const real_t y[neq], real_t dy[neq])
{
  // You may assume that NEQ == 1 for the test problem, but in general
  // NEQ refers to the Number of EQuations in the system of ODEs
}

int main(void)
{
   real_t tstart = 0.0;
   real_t tend = 2.0;
   size_t nsteps = 16;  // so that h=1/8
   real_t y[1];

   y[0] = 0.0;
   // other variables as needed
   // and call RK4

   printf("RK4 result: %g\n", y[0]);
   printf("truth: %g\n", exp(1.0));
}

File: Assignment-2/sep12/fortran/ode.f
!  MAIN - test program for the RK method.
!
      module ode
      use rkmod
      implicit none
      contains
!------------------------------------------------------------------------------
      subroutine testproblem(t, y, n, dy)
!  TESTPROBLEM - this routine's calling sequency must match that in the
!  abstract interface declared in rkmod.f.  The names of the dummy variables
!  need not be the same, but their type, kind, and rank must match.
!  Nere H is a
!
      integer, intent(in):: n  ! assumed 1 for testing
      real(WP), intent(in):: t, y(n)
      real(WP), intent(out):: dy(n)
!
!  your code here
!
      return
      end subroutine testproblem
!------------------------------------------------------------------------------
      end module ode
!------------------------------------------------------------------------------
      program test
!  TEST - test the RK solver
      use ode
      use, intrinsic:: iso_fortran_env
      implicit none
!
      real(WP):: tstart=0, tend=2, one=1, y(1)
      integer:: nsteps=16

      y(1) = 1.0
      call rk4(testproblem, y, 1, tstart, tend, nsteps)
      write(OUTPUT_UNIT,*) 'rk4:', y
      write(OUTPUT_UNIT,*) 'truth:', exp(one)
      stop
      end program test

